src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage_WriteBatching.cs

```diff
@@ -72,16 +73,38 @@ internal partial class SQLitePersistentStorage
             ArrayBuilder<Action<SqlConnection>> writesToProcess,
             CancellationToken cancellationToken)
         {
-            // Get the task that is responsible for doing the writes for this queue.
-            // This task will complete when all previously enqueued writes for this queue
-            // complete, and all the currently enqueued writes for this queue complete as well.
-            var writeTask = await GetWriteTaskAsync().ConfigureAwait(false);
-            await writeTask.ConfigureAwait(false);
+            // Get's the task representing the current writes being performed by another
+            // thread for this queue+key, and a TaskCompletionSource we can use to let
+            // other threads know about our own progress writing any new writes in this queue.
+            var (previousWritesTask, taskCompletionSource) = await GetWriteTaskAsync().ConfigureAwait(false);
+
+            // Wait for all previous writes to be flushed.
+            await previousWritesTask.ConfigureAwait(false);
+
+            if (writesToProcess.Count == 0)
+            {
+                // No additional writes for us to flush.  We can immediately bail out.
+                Debug.Assert(taskCompletionSource == null);
+                return;
+            }
+
+            // Now, if we have writes of our own, do them on this thread.
+            // 
+            // Note: this flushing is not cancellable.  We've already removed the
+            // writes from the write queue.  If we were not to write them out we
+            // would be losing data.
+            Debug.Assert(taskCompletionSource != null);
+
+            ProcessWriteQueue(connection, writesToProcess);
+
+            // Mark our TCS as completed.  Any other threads waiting on us will now be able
+            // to proceed.
+            taskCompletionSource.TrySetResult(0);
```

Comment:
interesting point.  i think technically it shoudl be in a try/catch/finally.   will update.